lib/gc/ExecutionEngine/GPURuntime/ocl/GpuOclRuntime.cpp

```diff
@@ -0,0 +1,690 @@
+    clBuildProgram(program, 1, &ctx->runtime.ext.device, nullptr, nullptr,
+                   nullptr);
+    CL_CHECKR(err, "Failed to build the program: ", program);
+    gcLogD("The program has been built: ", program);
+
+    auto kernel = clCreateKernel(program, name, &err);
+    CL_CHECKR(err, "Failed to create OpenCL kernel from program: ", program);
+    gcLogD("Created new OpenCL kernel ", kernel, " from program ", program);
+
+    cl_bool enable = CL_TRUE;
+    err = clSetKernelExecInfo(kernel,
+                              CL_KERNEL_EXEC_INFO_INDIRECT_HOST_ACCESS_INTEL,
+                              sizeof(enable), &enable);
+    CL_CHECKR(err, "Failed to set indirect host access.");
+    err = clSetKernelExecInfo(kernel,
+                              CL_KERNEL_EXEC_INFO_INDIRECT_DEVICE_ACCESS_INTEL,
+                              sizeof(enable), &enable);
+    CL_CHECKR(err, "Failed to set indirect device access.");
+    err = clSetKernelExecInfo(kernel,
+                              CL_KERNEL_EXEC_INFO_INDIRECT_SHARED_ACCESS_INTEL,
+                              sizeof(enable), &enable);
+    CL_CHECKR(err, "Failed to set indirect shared access.");
+
+    return new Kernel(program, kernel, gridSize, blockSize, argNum, argSize);
+  }
+
+  static void kernelDestroy(size_t count, Kernel **kernels) {
+    gcLogD("Destroying kernels.");
+    for (size_t i = 0; i < count; i++) {
+      if (kernels[i]) {
+        delete kernels[i];
+      }
+    }
+  }
+
+  static void kernelLaunch(OclContext *ctx, Kernel *kernel, ...) {
+    struct ClonedKernel {
+      cl_kernel kernel;
+
+      explicit ClonedKernel(cl_kernel kernel) : kernel(kernel) {}
+
+      ~ClonedKernel() {
+        gcLogD("Releasing cloned OpenCL kernel: ", kernel);
+        clReleaseKernel(kernel);
+      }
+    };
+
+    gcLogD("Launching kernel: ", kernel->kernel);
+
+    cl_int err;
+    ClonedKernel cloned{clCloneKernel(kernel->kernel, &err)};
+    CL_CHECKR(err, "Failed to clone OpenCL kernel: ", kernel->kernel);
+    gcLogD("Cloned OpenCL kernel ", kernel->kernel, ": ", cloned.kernel);
+
+    va_list args;
+    va_start(args, kernel);
+    for (size_t i = 0, n = kernel->argSize.size(); i < n; i++) {
+      auto size = kernel->argSize[i];
+      switch (size) {
+      case 0: {
+        void *ptr = va_arg(args, void *);
+        if (ctx->clPtrs.find(ptr) == ctx->clPtrs.end()) {
+          gcLogD("Setting kernel ", cloned.kernel, " argument ", i,
+                 " to USM pointer ", ptr);
+          err = ctx->runtime.ext.clSetKernelArgMemPointerINTEL(cloned.kernel, i,
+                                                               ptr);
+        } else {
+          gcLogD("Setting kernel ", cloned.kernel, " argument ", i,
+                 " to CL pointer ", ptr);
+          err = clSetKernelArg(cloned.kernel, i, sizeof(cl_mem), &ptr);
+        }
+        break;
+      }
+      case 1: {
+        auto val = static_cast<int8_t>(va_arg(args, int));
```

Comment:
?
```suggestion
        auto val = static_cast<int8_t>(va_arg(args, int8_t));
```

Reply:
No, the compiler warns that the arg is promoted to int and the behavior is not specified.